### DISCUSSION TITLE
Applying texture to flexcomp object

### DISCUSSION THREAD

**Discussion by sachinkundu** [google-deepmind/mujoco#2013] [Asking for Help]
## How to apply texture to a flexcomp object?

I have a simple flexcomp object defined like 
```xml
<mujoco model="cloth">
    <option timestep="0.01" integrator="implicitfast" viscosity="0.3" solver="CG" tolerance="1e-6"/>

    <extension>
        <plugin plugin="mujoco.elasticity.shell"/>
    </extension>

    <asset>
		<texture name="cloth_texture" type="2d" file="/home/sachin/source/thesis/assets/texture.png" />
		<material name="cloth_material" texture="cloth_texture" />
	</asset>

    <size nkey="10"/>
    <worldbody>
        <geom name="floor" type="plane" size="0 0 .1"/>
        <light diffuse=".6 .6 .6" specular="0.2 0.2 0.2" pos="0 0 4" dir="0 0 -1"/>
        <body>
            <body name="object" pos="0 0 0.9">
                <flexcomp type="grid" count="20 10 1" spacing=".0125 .0125 .0125" mass="0.5"
                          name="cloth" radius="0.006" material="cloth_material">
                    <!-- <contact condim="5" solref="0.05" friction="3"/> -->
                    <edge equality="true" damping="1.0"/>
                    <plugin plugin="mujoco.elasticity.shell">
                        <config key="poisson" value="0.4032740512363587"/>
                        <config key="thickness" value="0.0005"/>
                        <config key="young" value="85242.02369946481"/>
                    </plugin>
                </flexcomp>
                <site rgba="0 0 0 0" size="0.005" pos="0 0 -0.045" name="bottom_site"/>
                <site rgba="0 0 0 0" size="0.005" pos="0 0 0.03" name="top_site"/>
                <site rgba="0 0 0 0" size="0.005" pos="0.03 0.03 0" name="horizontal_radius_site"/>
            </body>
        </body>
    </worldbody>
</mujoco>
```
No problem loading the xml in `simulate` but the texture is not applied. The texture png is attached.

The documentation for [flexcomp](https://mujoco.readthedocs.io/en/stable/XMLreference.html#body-flexcomp) says that **Note that flexcomp does not generate texture coordinates automatically, except for 2D grids.** which this definition is. The documentation for [flex]() which the flexcomp should generate does say that **Texture coordinates for each vertex. If omitted, texture mapping for this flex is disabled, even if a texture is specified in the material.** which is what seems to be happening.

Is there any example somewhere how to apply a texture to a flexcomp object?
![texture](https://github.com/google-deepmind/mujoco/assets/233623/ae1d9e6c-cc11-4ec6-9f2b-7bf9cbb62389)

**Comment by kevinzakka** [comment]
cc @quagla

**Comment by quagla** [comment]
Hi, I have just found a bug related to this. I am pushing out a fix soon.

**Reply by harshaguda** [reply]
Hi @quagla, is the bug resolved? I could not find any issues related to this.

**Reply by quagla** [reply]
Yes it was fixed in 831d9881d3c2e453536425c5b4eeb65c1aeff464 I mentioned this discussion in the commit message but unfortunately discussions don't get closed automatically like issues do.

**Comment by jsll** [comment]
@quagla I still don't think this issue is solved. At least not for type=grid.

Below is a minimal example of two different object using the same texture: a box and a grid.

```
<mujoco model="Soft box">
  <include file="scene.xml"/>

  <option solver="CG" tolerance="1e-6"/>

  <size memory="30M"/>

  <worldbody>
    <body pos="0 0 1" name="body">
      <freejoint/>
      <geom size=".1" contype="0" conaffinity="0" group="4"/>
      <flexcomp name="box" type="box" count="7 7 2" spacing=".04 .04 .04"
                radius="0.01" dim="3" material="matsponge">
        <contact internal="false" selfcollide="none"/>
        <edge equality="true"/>
      </flexcomp>
    </body>
    <body pos="0.4 0 1" name="cloth">
      <freejoint/>
      <geom size=".1" contype="0" conaffinity="0" group="4"/>
      <flexcomp name="cloth" type="grid" count="7 7 1" spacing=".04 .04 .04"
                radius="0.01" dim="2" material="matsponge">
        <contact internal="false" selfcollide="none"/>
        <edge equality="true"/>
      </flexcomp>
    </body>
  </worldbody>
</mujoco>
```

As can clearly be seen in the figure below the box object get the texture applied while the gird object does not
![image](https://github.com/user-attachments/assets/599ca300-ca35-49cf-9ddd-d825f436eef6)

**Reply by quagla** [reply]
Let me know if 6a862478103ee7c618b2a1f9b84c6f21fcc5d882 fixes it.

**Reply by jsll** [reply]
@quagla As shown in the image below, https://github.com/google-deepmind/mujoco/commit/6a862478103ee7c618b2a1f9b84c6f21fcc5d882 
fixes the issue. Thank you so much! :+1: 

![image](https://github.com/user-attachments/assets/2c29f8fd-7662-4533-b235-45b73337cd9a)
